### PR TITLE
Fix crash in git_clone on extremely large repos

### PR DESCRIPTION
**Pull request by jeffhostetler**
A crash was observed on Windows when cloning a very large repo. In this instance, the received packfile was nearly 6Gb.

The problem was in sloppy casting of the "size" field from src/indexer.c:append_to_pack() to src/win32/posix_w32.c:p_ftruncate() to _chsize_s().
Where the file position was being truncated and then sign-extended.

Also fixed the return value from _chsize_s() which differs from _chsize().

### CODE REVIEW

**Comment by jeffhostetler** [comment]
Looks like this breaks ming-win32 which (still) doesn't have _chsize_s().  Need to investigate this a bit.

**Comment by jeffhostetler** [comment]
I re-worked the change in p_ftruncate() for Windows and restored the original code calling _chsize() which appears to only be needed by ming-w32.  

However, I did add argument checking. 

I'm wondering if we want to have simple clar test that just calls p_ftruncate(fd, 2Gb+1 or maybe 4Gb+1) to confirm this in isolation.

**Comment by ethomson** [comment]
That'd be great; it would be nice to have a test that tests the problem and not something that depends on a big repo.  We have some tests that write big files (defaulting to off), looking at the `GITTEST_INVASIVE_FS_SIZE` environment variable.  (One of the filter tests.)

File: src/win32/posix_w32.c [hunk 1]
```diff
@@ -42,12 +42,26 @@
...
 #if defined(_MSC_VER) && _MSC_VER >= 1500
-	return _chsize_s(fd, size);
+	return ((_chsize_s(fd, size) == 0) ? 0 : -1);
 #else
-	return _chsize(fd, size);
+	/* TODO Find a replacement for _chsize() that handles big files.
+	 *      This comment is probably __MINGW32__ specific.
+	 */
+	if (size > 0x7fffffff)
+		return -1;
```
Comment (carlosmn) [inline src/win32/posix_w32.c]: The caller would inspect `errno` if this function fails, no? We should set it to something useful

**Comment by carlosmn** [comment]
Yeah, let's go for `GITTEST_INVASIVE_FS_SIZE` and create one or two large objects from a random source which we can then clone locally

File: src/win32/posix_w32.c (hunk 1 above)
Comment (jeffhostetler) [inline src/win32/posix_w32.c]: good catch. i'll set errno to something sane like the actual _chsize() would.

**Comment by jeffhostetler** [comment]
I've added a simple "open(); ftruncate(); lstat()" unit test to verify that the ftruncate() works as expected for big files.

**Comment by jeffhostetler** [comment]
I'm adding another test to "clone empty_test_repo", stage and commit a huge file, and locally clone it
but ran into https://github.com/libgit2/libgit2/issues/2936

I think I'll commit my simple unit test for ftruncate and let us close this problem and not look for a solution to the blog problem in this PR.

File: tests/core/ftruncate.c [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+	p_unlink(filename);
+}
+
+static void _extend(git_off_t i64len)
+{
+	struct stat st;
+	int error;
+
+	assert(sizeof(i64len) == 8); /* yes, i'm that paranoid. */
```
Comment (carlosmn) [inline tests/core/ftruncate.c]: We do guarantee this through the headers (which is why we don't use `off_t`).

File: src/win32/posix_w32.c [hunk 3]
```diff
@@ -42,12 +42,30 @@
...
+
 #if defined(_MSC_VER) && _MSC_VER >= 1500
-	return _chsize_s(fd, size);
+	return ((_chsize_s(fd, size) == 0) ? 0 : -1);
 #else
-	return _chsize(fd, size);
+	/* TODO Find a replacement for _chsize() that handles big files.
+	 *      This comment is probably __MINGW32__ specific.
+	 */
+	if (size > 0x7fffffff) {
```
Comment (carlosmn) [inline src/win32/posix_w32.c]: We should have `INT32_MAX` defined in any OS, it'd be more descriptive to use that name.

**Comment by jeffhostetler** [comment]
OK. I fixed the constants and squashed everything.  With that it now just looks like a fix to p_ftruncate() with nothing to indicate it was discovered via git_clone().  So I'd like to go ahead and mark this finished.  If/when I get time, I'll do another PR containing a unit test to build upon this and try to clone a huge repo.

File: src/win32/posix_w32.c [hunk 4]
```diff
@@ -42,12 +42,30 @@
...
+ * files may be longer than 2Gb.
+ */
+int p_ftruncate(int fd, git_off_t size)
 {
+	if (size < 0) {
+		errno = EINVAL;
+		return -1;
+	}
+
 #if defined(_MSC_VER) && _MSC_VER >= 1500
```
Comment (ethomson) [inline src/win32/posix_w32.c]: I think that this (historic) test is incorrect.  `_chsize_s` was available way back in XP.  Should this just be `#ifdef __MINGW32__` ?

**Comment by carlosmn** [comment]
Agreed, let's just go with the truncate fix. We can have some PR creating three or four 2GB objects which we can then clone, but allowing the creation of an object larger than `SIZE_MAX` is going to require quite some changes.

**Pull request merged**